crates/matrix-sdk/src/event_cache/store.rs

```diff
@@ -31,40 +31,173 @@ pub trait EventCacheStore: Send + Sync {
     /// Returns all the known events for the given room.
     async fn room_events(&self, room: &RoomId) -> Result<Vec<SyncTimelineEvent>>;
 
-    /// Adds all the events to the given room.
-    async fn add_room_events(&self, room: &RoomId, events: Vec<SyncTimelineEvent>) -> Result<()>;
+    /// Adds all the entries to the given room's timeline.
+    async fn append_room_entries(&self, room: &RoomId, entries: Vec<TimelineEntry>) -> Result<()>;
 
-    /// Clear all the events from the given room.
-    async fn clear_room_events(&self, room: &RoomId) -> Result<()>;
+    /// Returns whether the store knows about the given pagination token.
+    async fn contains_gap(&self, room: &RoomId, pagination_token: &PaginationToken)
+        -> Result<bool>;
+
+    /// Replaces a given gap (identified by its pagination token) with the given
+    /// entries.
+    ///
+    /// Note: if the gap hasn't been found, then nothing happens, and the events
+    /// are lost.
+    ///
+    /// Returns whether the gap was found.
+    async fn replace_gap(
+        &self,
+        room: &RoomId,
+        gap_id: Option<&PaginationToken>,
+        entries: Vec<TimelineEntry>,
+    ) -> Result<bool>;
+
+    /// Retrieve the earliest (oldest backpagination token for the given room.)
+    async fn earliest_backpagination_token(&self, room: &RoomId)
+        -> Result<Option<PaginationToken>>;
+
+    /// Clear all the information tied to a given room.
+    ///
+    /// This forgets the following:
+    /// - events in the room
+    /// - pagination tokens
+    async fn clear_room(&self, room: &RoomId) -> Result<()>;
+}
+
+/// A newtype wrapper for a pagination token returned by a /messages response.
+#[derive(Clone, Debug, PartialEq)]
+pub struct PaginationToken(pub String);
+
+#[derive(Clone)]
+pub enum TimelineEntry {
```

Comment:
This type is going to be removed once https://github.com/matrix-org/matrix-rust-sdk/pull/3166 is merged: we will need to reconciliate those.